### PR TITLE
add planWithEndEffectorTwist

### PR DESCRIPTION
**Pull request by nansongyi**
This PR adds `planWithEndEffectorTwist`

[Explain how this pull request was tested.]

***

**Before creating a pull request**

- [x] Document new methods and classes
- [ ] Format code with `make format`

**Before merging a pull request**

- [ ] Set version target by selecting a milestone on the right side
- [ ] Summarize this change in `CHANGELOG.md`
- [ ] Add unit test(s) for this change

### CODE REVIEW

**Comment by brianhou** [comment]
@gilwoolee can you review this PR? Thanks!

**Comment by gilwoolee** [comment]
@nansongyi could you resolve the conflict?

File: include/aikido/constraint/FiniteSampleable.hpp [hunk 1]
```diff
@@ -17,15 +18,15 @@ class FiniteSampleable : public Sampleable
   /// \param _state The only sample in this constraint.
   FiniteSampleable(
       statespace::StateSpacePtr _stateSpace,
-      const statespace::StateSpace::State* _state);
+      const aikido::statespace::dart::MetaSkeletonStateSpace::ScopedState& _state);
```
Comment (gilwoolee) [inline include/aikido/constraint/FiniteSampleable.hpp]: Please revert this to the original. This class is above `aikido/statespace/dart` so it shouldn't take `dart::MetaSkeletonStateSpace` component as its parameter.

File: include/aikido/constraint/FiniteSampleable.hpp [hunk 2]
```diff
@@ -17,15 +18,15 @@ class FiniteSampleable : public Sampleable
...
 
   /// Constructor for multiple samples.
   /// \param _stateSpace StateSpace in which _states belong.
   /// \param _states Samples in this constraint.
   ///        SampleGenerator will generate samples in this order.
   FiniteSampleable(
       statespace::StateSpacePtr _stateSpace,
-      const std::vector<const statespace::StateSpace::State*>& _states);
+      const std::vector<aikido::statespace::dart::MetaSkeletonStateSpace::ScopedState>& _states);
```
Comment (gilwoolee) [inline include/aikido/constraint/FiniteSampleable.hpp]: Same as above.

File: include/aikido/constraint/FiniteSampleable.hpp [hunk 3]
```diff
@@ -43,7 +44,7 @@ class FiniteSampleable : public Sampleable
 
 private:
   statespace::ConstStateSpacePtr mStateSpace;
-  std::vector<statespace::StateSpace::State*> mStates;
+  std::vector<aikido::statespace::dart::MetaSkeletonStateSpace::ScopedState> mStates;
```
Comment (gilwoolee) [inline include/aikido/constraint/FiniteSampleable.hpp]: Same as above.

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp [hunk 4]
```diff
@@ -7,6 +7,10 @@
 #include <aikido/statespace/dart/MetaSkeletonStateSpace.hpp>
 #include <aikido/trajectory/Interpolated.hpp>
 
+#include "aikido/trajectory/Spline.hpp"
```
Comment (gilwoolee) [inline include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: Follow the convention above to use `< ... >`.

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp [hunk 5]
```diff
@@ -7,6 +7,10 @@
...
+// #include "aikido/robot/util.hpp"
```
Comment (gilwoolee) [inline include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: Remove these dead lines.

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp [hunk 6]
```diff
@@ -97,6 +101,7 @@ aikido::trajectory::UniqueInterpolatedPtr planToEndEffectorOffset(
 /// \return Trajectory or \c nullptr if planning failed.
 aikido::trajectory::UniqueInterpolatedPtr planToEndEffectorPose(
     const aikido::statespace::dart::MetaSkeletonStateSpacePtr& stateSpace,
+    const statespace::dart::MetaSkeletonStateSpace::State& startState,
```
Comment (gilwoolee) [inline include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: Add docstring for this parameter below line 85.

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp [hunk 7]
```diff
@@ -109,6 +114,23 @@ aikido::trajectory::UniqueInterpolatedPtr planToEndEffectorPose(
     std::chrono::duration<double> timelimit,
     planner::Planner::Result* result = nullptr);
 
+// std::unique_ptr<aikido::trajectory::Spline> planWithEndEffectorTwist(
```
Comment (gilwoolee) [inline include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: Remove this line.

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp (hunk 6 above)
Comment (gilwoolee) [reply include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: Actually, why does this method need to change in this PR? Please make a separate PR for this and have this PR be only about `planWithEndEffectorTwist`

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp [hunk 8]
```diff
@@ -109,6 +114,23 @@ aikido::trajectory::UniqueInterpolatedPtr planToEndEffectorPose(
...
+aikido::trajectory::UniqueInterpolatedPtr planWithEndEffectorTwist(
```
Comment (gilwoolee) [inline include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: Need a full docstring for this method.

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp [hunk 9]
```diff
@@ -109,6 +114,23 @@ aikido::trajectory::UniqueInterpolatedPtr planToEndEffectorPose(
...
+    const aikido::statespace::dart::ConstMetaSkeletonStateSpacePtr& stateSpace,
+    const statespace::dart::MetaSkeletonStateSpace::State& startState,
+    ::dart::dynamics::MetaSkeletonPtr metaskeleton,
+    const ::dart::dynamics::ConstBodyNodePtr& bn,
+    const Eigen::Vector6d& twistSeq,
+    double durationSeq,
+    const aikido::constraint::ConstTestablePtr& constraint,
```
Comment (gilwoolee) [inline include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: I'd prefer having this before `twistSeq`.

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp [hunk 10]
```diff
@@ -109,6 +114,23 @@ aikido::trajectory::UniqueInterpolatedPtr planToEndEffectorPose(
     std::chrono::duration<double> timelimit,
     planner::Planner::Result* result = nullptr);
 
+// std::unique_ptr<aikido::trajectory::Spline> planWithEndEffectorTwist(
+aikido::trajectory::UniqueInterpolatedPtr planWithEndEffectorTwist(
+    const aikido::statespace::dart::ConstMetaSkeletonStateSpacePtr& stateSpace,
+    const statespace::dart::MetaSkeletonStateSpace::State& startState,
+    ::dart::dynamics::MetaSkeletonPtr metaskeleton,
+    const ::dart::dynamics::ConstBodyNodePtr& bn,
+    const Eigen::Vector6d& twistSeq,
```
Comment (gilwoolee) [inline include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: Let's use a full variable name, e.g. twistSequence? (I'm actually not sure what `twistSeq` stands for.)

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp [hunk 11]
```diff
@@ -109,6 +114,23 @@ aikido::trajectory::UniqueInterpolatedPtr planToEndEffectorPose(
...
+    double durationSeq,
```
Comment (gilwoolee) [inline include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: Same as above.

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp [hunk 12]
```diff
@@ -109,6 +114,23 @@ aikido::trajectory::UniqueInterpolatedPtr planToEndEffectorPose(
...
+    const aikido::constraint::ConstTestablePtr& constraint,
+    double positionTolerance,
+    double angularTolerance,
+    double initialStepSize,
+    double jointLimitTolerance,
+    double constraintCheckResolution,
+    std::chrono::duration<double> timelimit,
+    planner::Planner::Result* result = nullptr);
```
Comment (gilwoolee) [inline include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: Use full path `aikido::planner::Planner::Result`

File: include/aikido/robot/ConcreteManipulator.hpp [hunk 13]
```diff
@@ -114,6 +114,32 @@ class ConcreteManipulator : public Manipulator
       double positionTolerance,
       double angularTolerance);
 
+  trajectory::TrajectoryPtr planToEndEffectorOffset(
```
Comment (gilwoolee) [inline include/aikido/robot/ConcreteManipulator.hpp]: Let's defer these changes on `ConcreteManpulator` class to a separate PR.

File: src/constraint/FiniteSampleable.cpp [hunk 14]
```diff
@@ -12,7 +12,7 @@ class FiniteSampleGenerator : public SampleGenerator
   // For internal use only.
   FiniteSampleGenerator(
       statespace::ConstStateSpacePtr _stateSpace,
-      const std::vector<statespace::StateSpace::State*>& _states);
+      const std::vector<aikido::statespace::dart::MetaSkeletonStateSpace::ScopedState>& _states);
```
Comment (gilwoolee) [inline src/constraint/FiniteSampleable.cpp]: Revert these, as I mentioned in the header file.

File: src/constraint/FiniteSampleable.cpp [hunk 15]
```diff
@@ -45,8 +45,9 @@ class FiniteSampleGenerator : public SampleGenerator
 //==============================================================================
 FiniteSampleGenerator::FiniteSampleGenerator(
     statespace::ConstStateSpacePtr _stateSpace,
-    const std::vector<statespace::StateSpace::State*>& _states)
-  : mStateSpace(std::move(_stateSpace)), mIndex(0)
+    const std::vector<aikido::statespace::dart::MetaSkeletonStateSpace::ScopedState>& _states)
```
Comment (gilwoolee) [inline src/constraint/FiniteSampleable.cpp]: I think all of these changes are irrelevant to this PR and should be removed.

File: src/constraint/SequentialSampleable.cpp [hunk 16]
```diff
@@ -89,7 +89,12 @@ bool SequentialSampleGenerator::sample(statespace::StateSpace::State* state)
   while (mIndex < mGenerators.size())
   {
     if (mGenerators[mIndex]->sample(state))
+    {
```
Comment (gilwoolee) [inline src/constraint/SequentialSampleable.cpp]: Same comment as above. Revert these.

File: src/constraint/dart/CollisionFree.cpp [hunk 17]
```diff
@@ -37,13 +37,19 @@ bool CollisionFree::isSatisfied(
...
     TestableOutcome* outcome) const
 {
-  auto collisionFreeOutcome
-      = dynamic_cast_or_throw<CollisionFreeOutcome>(outcome);
-
-  if (collisionFreeOutcome)
-  {
-    collisionFreeOutcome->clear();
-  }
+  // CollisionFreeOutcome* collisionFreeOutcome
```
Comment (gilwoolee) [inline src/constraint/dart/CollisionFree.cpp]: Same as above.

File: src/control/KinematicSimulationTrajectoryExecutor.cpp [hunk 18]
```diff
@@ -139,6 +139,7 @@ void KinematicSimulationTrajectoryExecutor::step(
   // Check if trajectory has completed.
   if (executionTime >= mTraj->getEndTime())
   {
+    std::cout << "Trajectory completed" << std::endl;
```
Comment (gilwoolee) [inline src/control/KinematicSimulationTrajectoryExecutor.cpp]: Same as above.

File: include/aikido/planner/vectorfield/VectorFieldPlanner.hpp (hunk 10 above)
Comment (nansongyi) [reply include/aikido/planner/vectorfield/VectorFieldPlanner.hpp]: I just followed the convention in `MoveEndEffectorTwistVectorField.cpp`.  I think twistSeq is just simply a twist msg(same in ROS), first 3 elements: angular velocity, last 3 elements: linear velocity.

**Comment by brianhou** [comment]
Closing in favor of the more isolated changes in #545.

**Pull request closed without merging**